File: packages/cli/src/scaffold.ts
import { join } from "node:path";
import { TerminalLogger } from "@talosjs/logger";
import { toPascalCase } from "@talosjs/utils/toPascalCase";
import { askConfirm } from "./prompts/askConfirm";
import { askName } from "./prompts/askName";
import { ensureModule, LOG_OPTIONS } from "./utils";

export type ScaffoldConfigType = {
  /** Human-readable resource label used in messages (e.g. "Cache") */
  label: string;
  /** Prompt shown when no name option is provided */
  promptMessage: string;
  /** Class and file name suffix appended to the normalized name (e.g. "Cache") */
  suffix: string;
  /** Suffixes stripped from the user-provided name; defaults to [suffix] */
  stripSuffixes?: string[];
  /** Source file template ({{NAME}} is replaced with the normalized name) */
  template: string;
  /** Test file template ({{NAME}} and {{MODULE}} are replaced) */
  testTemplate: string;
  /** Subdirectory under the module's src/ (e.g. "cache") */
  dir: string;
  /** Subdirectory under the module's tests/; defaults to dir */
  testsDir?: string;
  /** Module class array to register the generated class into (e.g. "cronJobs") */
  moduleField?: string;
  /** Runtime package installed with `bun add` when missing (e.g. "@talosjs/cache") */
  dependency?: string;
  /** Extra {{KEY}} replacements applied to the source template */
  templateData?: (name: string) => Record<string, string>;
  /** Static sibling files written into the resource dir once, keyed by file name (skipped if they already exist) */
  staticFiles?: Record<string, string>;
};

export type ScaffoldOptionsType = {
  name?: string;
  module?: string;
  override?: boolean;
};

export const addClassToModule = async (
  modulePath: string,
  className: string,
  importDir: string,
  field: string,
): Promise<void> => {
  let content = await Bun.file(modulePath).text();
  const importLine = `import { ${className} } from "./${importDir}/${className}";\n`;

  const lastImportIndex = content.lastIndexOf("import ");
  const lineEnd = content.indexOf("\n", lastImportIndex);
  content = `${content.slice(0, lineEnd + 1)}${importLine}${content.slice(lineEnd + 1)}`;

  const regex = new RegExp(`(${field}:\\s*\\[)([^\\]]*)`, "s");
  const match = content.match(regex);
  if (match) {
    const existing = match[2]?.trim();
    const newValue = existing ? `${existing}, ${className}` : className;
    content = content.replace(regex, `$1${newValue}`);
  }

  await Bun.write(modulePath, content);
};

export const installDependency = async (dependency: string): Promise<void> => {
  const packageJsonPath = join(process.cwd(), "package.json");
  const packageJson = await Bun.file(packageJsonPath).json();
  const deps = packageJson.dependencies ?? {};
  const devDeps = packageJson.devDependencies ?? {};

  if (!deps[dependency] && !devDeps[dependency]) {
    const install = Bun.spawn(["bun", "add", dependency], {
      cwd: process.cwd(),
      stdout: "ignore",
      stderr: "inherit",
    });
    await install.exited;
  }
};

export const scaffoldResource = async (config: ScaffoldConfigType, options: ScaffoldOptionsType): Promise<void> => {
  let { name, module = "shared", override = false } = options;

  if (!name) {
    name = await askName({ message: config.promptMessage });
  }

  name = toPascalCase(name);
  for (const suffix of config.stripSuffixes ?? [config.suffix]) {
    name = name.replace(new RegExp(`${suffix}$`), "");
  }

  const content = renderTemplate(config, name);

  await ensureModule(module);

  const base = join("modules", module);
  const localDir = join(base, "src", config.dir);
  const fileName = `${name}${config.suffix}.ts`;
  const filePath = join(process.cwd(), localDir, fileName);

  if (!override && (await Bun.file(filePath).exists())) {
    const shouldOverride = await askConfirm({
      message: `${config.label} "${name}${config.suffix}" already exists. Override it?`,
      initial: false,
    });

    if (!shouldOverride) {
      return;
    }
  }

  await Bun.write(filePath, content);

  // Generate test file
  const testContent = config.testTemplate.replace(/{{NAME}}/g, name).replace(/{{MODULE}}/g, module);
  const testsLocalDir = join(base, "tests", config.testsDir ?? config.dir);
  const testFilePath = join(process.cwd(), testsLocalDir, `${name}${config.suffix}.spec.ts`);
  await Bun.write(testFilePath, testContent);

  // Register the generated class in its module
  if (config.moduleField) {
    const modulePascalName = toPascalCase(module);
    const modulePath = join(process.cwd(), base, "src", `${modulePascalName}Module.ts`);
    if (await Bun.file(modulePath).exists()) {
      await addClassToModule(modulePath, `${name}${config.suffix}`, config.dir, config.moduleField);
    }
  }

  const logger = new TerminalLogger();

  logger.success(`${join(localDir, name)}${config.suffix}.ts created successfully`, undefined, LOG_OPTIONS);

  logger.success(`${join(testsLocalDir, name)}${config.suffix}.spec.ts created successfully`, undefined, LOG_OPTIONS);

  // Write static sibling files (e.g. a shared dictionary) once, without overriding existing ones
  for (const [staticName, staticContent] of Object.entries(config.staticFiles ?? {})) {
    const staticPath = join(process.cwd(), localDir, staticName);
    if (!(await Bun.file(staticPath).exists())) {
      await Bun.write(staticPath, staticContent);
      logger.success(`${join(localDir, staticName)} created successfully`, undefined, LOG_OPTIONS);
    }
  }

  // Install the runtime dependency if not already installed
  if (config.dependency) {
    await installDependency(config.dependency);
  }
};

const renderTemplate = (config: ScaffoldConfigType, name: string): string => {
  let content = config.template.replace(/{{NAME}}/g, name);

  for (const [key, value] of Object.entries(config.templateData?.(name) ?? {})) {
    content = content.replace(new RegExp(`{{${key}}}`, "g"), value);
  }

  return content;
};

File: packages/rust-cli/src/utils/scaffold.rs
use std::fs;
use std::path::PathBuf;
use std::process::Command;

use regex::Regex;

use super::case::to_pascal_case;
use super::process::run_spinner_step;
use super::prompts::{ask_confirm, ask_input};

#[derive(Default)]
pub struct ScaffoldConfig {
    pub label: &'static str,
    pub prompt_message: &'static str,
    pub suffix: &'static str,
    pub template: &'static str,
    pub test_template: &'static str,
    pub dir: &'static str,
    pub tests_dir: Option<&'static str>,
    pub strip_suffixes: &'static [&'static str],
    pub module_field: Option<&'static str>,
    pub dependency: Option<&'static str>,
    pub template_data: Option<TemplateDataFn>,
}

pub type TemplateDataFn = Box<dyn Fn(&str) -> Vec<(&'static str, String)>>;

pub struct ScaffoldOptions {
    pub name: Option<String>,
    pub module: Option<String>,
    pub r#override: bool,
}

pub fn ensure_module(module: &str, cwd: &std::path::Path) {
    let package_json = cwd.join("modules").join(module).join("package.json");
    if package_json.exists() {
        return;
    }
    crate::commands::module_create::execute(crate::commands::module_create::ModuleCreateOptions {
        name: module.to_string(),
        destination: Some("app".to_string()),
        cwd: cwd.to_path_buf(),
        silent: true,
    });
}

pub fn add_class_to_module(
    module_path: &std::path::Path,
    class_name: &str,
    import_dir: &str,
    field: &str,
) -> Result<(), String> {
    let mut content = fs::read_to_string(module_path).map_err(|e| e.to_string())?;
    let import_line = format!("import {{ {class_name} }} from \"./{import_dir}/{class_name}\";\n");

    let last_import_index = content.rfind("import ").unwrap_or(0);
    let line_end = content[last_import_index..]
        .find('\n')
        .map(|i| last_import_index + i)
        .unwrap_or(content.len().saturating_sub(1));
    content = format!(
        "{}{import_line}{}",
        &content[..=line_end.min(content.len().saturating_sub(1))],
        &content[(line_end + 1).min(content.len())..]
    );

    if let Ok(re) = Regex::new(&format!(r"(?s)({field}:\s*\[)([^\]]*)"))
        && let Some(caps) = re.captures(&content)
    {
        let existing = caps.get(2).map(|m| m.as_str().trim()).unwrap_or("");
        let new_value = if existing.is_empty() {
            class_name.to_string()
        } else {
            format!("{existing}, {class_name}")
        };
        let whole = caps.get(0).unwrap().as_str().to_string();
        let prefix = caps.get(1).unwrap().as_str().to_string();
        content = content.replacen(&whole, &format!("{prefix}{new_value}"), 1);
    }

    fs::write(module_path, content).map_err(|e| e.to_string())
}

pub fn install_dependency(dependency: &str, cwd: &std::path::Path) {
    let package_json_path = cwd.join("package.json");
    let Ok(raw) = fs::read_to_string(&package_json_path) else {
        return;
    };
    let Ok(package_json) = serde_json::from_str::<serde_json::Value>(&raw) else {
        return;
    };
    let already_present = ["dependencies", "devDependencies"].iter().any(|key| {
        package_json
            .get(key)
            .and_then(|deps| deps.get(dependency))
            .is_some()
    });
    if already_present {
        return;
    }

    run_spinner_step(
        false,
        &format!("Installing {dependency}"),
        Command::new("bun")
            .args(["add", dependency])
            .current_dir(cwd),
    );
}

pub fn scaffold_resource(config: &ScaffoldConfig, options: ScaffoldOptions, cwd: &std::path::Path) {
    let name = match options.name {
        Some(name) => name,
        None => match ask_input(config.prompt_message) {
            Some(name) => name,
            None => return,
        },
    };

    let module = options.module.unwrap_or_else(|| "shared".to_string());

    let mut name = to_pascal_case(&name);
    let strip_suffixes: &[&str] = if config.strip_suffixes.is_empty() {
        std::slice::from_ref(&config.suffix)
    } else {
        config.strip_suffixes
    };
    for suffix in strip_suffixes {
        if let Some(stripped) = name.strip_suffix(suffix) {
            name = stripped.to_string();
        }
    }

    let mut content = config.template.replace("{{NAME}}", &name);
    if let Some(template_data) = &config.template_data {
        for (key, value) in template_data(&name) {
            content = content.replace(&format!("{{{{{key}}}}}"), &value);
        }
    }

    ensure_module(&module, cwd);

    let base = cwd.join("modules").join(&module);
    let local_dir = base.join("src").join(config.dir);
    let file_name = format!("{name}{}.ts", config.suffix);
    let file_path = local_dir.join(&file_name);

    if !options.r#override
        && file_path.exists()
        && !ask_confirm(
            &format!(
                "{} \"{name}{}\" already exists. Override it?",
                config.label, config.suffix
            ),
            false,
        )
    {
        return;
    }

    if let Some(parent) = file_path.parent() {
        let _ = fs::create_dir_all(parent);
    }
    if fs::write(&file_path, content).is_err() {
        super::style::error(format!("Failed to write {}", file_path.display()));
        return;
    }

    let test_content = config
        .test_template
        .replace("{{NAME}}", &name)
        .replace("{{MODULE}}", &module);
    let tests_local_dir = base
        .join("tests")
        .join(config.tests_dir.unwrap_or(config.dir));
    let test_file_path = tests_local_dir.join(format!("{name}{}.spec.ts", config.suffix));
    if let Some(parent) = test_file_path.parent() {
        let _ = fs::create_dir_all(parent);
    }
    let _ = fs::write(&test_file_path, test_content);

    if let Some(module_field) = config.module_field {
        let module_pascal_name = to_pascal_case(&module);
        let module_path = base
            .join("src")
            .join(format!("{module_pascal_name}Module.ts"));
        if module_path.exists() {
            let _ = add_class_to_module(
                &module_path,
                &format!("{name}{}", config.suffix),
                config.dir,
                module_field,
            );
        }
    }

    super::style::success(format!(
        "{} created successfully",
        local_dir
            .join(format!("{name}{}.ts", config.suffix))
            .display()
    ));
    super::style::success(format!(
        "{} created successfully",
        tests_local_dir
            .join(format!("{name}{}.spec.ts", config.suffix))
            .display()
    ));

    if let Some(dependency) = config.dependency {
        install_dependency(dependency, cwd);
    }
}

pub fn current_dir() -> PathBuf {
    std::env::current_dir().unwrap_or_default()
}
